### PR TITLE
Feature/enhance admin sponsors

### LINKED ISSUE #1375: Add more information to sponsored "Sponsored section" in their admin area
## Description of the issue 📄

Currently on the codebar website we list all the times a company has hosted a workshop in their admin area. However it only lists workshops, I also want to know if they've hosted a monthlies event or been a host/sponsor of a one-off event.

## Screenshots 📷

screencapture-codebar-io-admin-sponsors-45-2020-08-04-15_16_26

## To do 📋

* [ ] Claim this issue (comment below, or assign yourself if you are part of the codebar org)
* [ ] Fork and clone the repository
* [ ] Update the relevant files. Follow the steps to fix section in this issue.
* [ ] Commit your changes as one commit. Use the title of this issue as your commit message
* [ ] Submit a pull request
* [ ] Mention this issue in the PR description by including it's number
* [ ] Have your pull request reviewed & merged by a codebar team member

### PR DESCRIPTION
**Pull request by despo**
- [x] Moves sponsor contacts to its own model - this enables a sponsor to have multiple contacts (current a contact is fields on Sponsor model) 
- [x] Adds migration to move data from Sponsor to Contacts model
- [x] Removes Sponsor - Member association from form and adds migration to move existing member data to Contacts model
(Future work after this has been deployed: Remove member and related associations from Sponsor model)
```
rake sponsors:contacts:migrate                         # Migrates contact information from Sponsor model to Contact model
rake sponsors:members:migrate_to_contacts              # Creates Contact entries for Sponsor associated Members
```
**Note**: These two rake tasks must be executed during the deployment of this functionality for the existing data to be migrated.

- [x] Update Sponsor form to support new contact functionality 
edit-sponsor

- [x] Update Sponsor view to display contacts and other key details (contacts, map preview, associated sponsorships etc)
view-sponsor

- [x] Adds sponsor contacts to the mailchimp Sponsors mailing list if they opt in and send them an email enabling to manage their contact preferences

### Sponsor mailing list subscription notification email
sponsor-contact-opt-in-email

### Sponsor contact's contact preferences pge
sponsor-contact-manage-contact-preferences

- [x] Improves admin/sponsors view
Screen Shot 2020-10-22 at 10 14 07 1

- [x] Adds view for sponsor contacts (in admin/contacts)
Screen Shot 2020-10-22 at 10 15 04 1

## Other things to note:
- Added sample pagination CSS from http://mislav.github.io/will_paginate/ 
- Used `cocoon` gem to manage nested dynamic forms. I was initially considering following [this guide](https://www.pluralsight.com/guides/ruby-on-rails-nested-attributes) but this is exactly what cocoon does so it made sense to use it instead

### Before
Create a new *Sponsor contacts* audience in Mailchimp and assign its id to the`SPONSOR_NEWSLETTER_ID`variable in production
`heroku config:set SPONSOR_NEWSLETTER_ID=XXXXXXX --app codebar-production`

### After
**Run the following rake tasks to migrate existing data**
rake sponsors:contacts:migrate                   
rake sponsors:members:migrate_to_contacts 

Resolves #1375 

### Additional changes

Log admin subscribing/unsubscribing contact to mailing list + contact subscribing/unsubscribing themselves. This is to prevent any miscommunication and re-adding contacts that have opted out on their own. 
I am also logging the email as text as I don't want to rely on the model data in case the email is edited and it no longer matches what is in mailchimp.
sponsor_contact_audit

### CODE REVIEW

**Review by KimberleyCook — COMMENTED** [review]
Designs look good, I like the addition of the map on the admin area of the sponsor.

I'll leave someone else to review to code part.

Nice work :)

File: app/controllers/admin/sponsors_controller.rb [hunk 1]
```diff
@@ -19,38 +22,51 @@ def create
...
+    flash[:warning] = @sponsor.errors.full_messages
+    render 'edit'
   end
 
   private
 
   def sponsor_params
     params.require(:sponsor).permit(:name, :avatar, :website, :seats, :accessibility_info,
                                     :number_of_coaches, :level, :email, :contact_first_name,
-                                    :contact_surname, contact_ids: [], address_attributes:
```
Comment (matyikriszta) [inline app/controllers/admin/sponsors_controller.rb:48]: @despo am I correct in thinking the `contact_first_name` & `contact_surname` fields are going to be removed/have been removed from Sponsor in another PR? I think you mentioned something to this effect in the PR description but wanted to double check.

File: app/controllers/contact_preferences_controller.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
+class ContactPreferencesController < ApplicationController
+  def show
+    @contact = Contact.find_by(token: token)
+
+    return if @contact
+
+    flash[:notice] = 'Invalid session'
```
Comment (matyikriszta) [inline app/controllers/contact_preferences_controller.rb:7]: @despo any reason why this string is not in the locale files?

File: app/controllers/admin/sponsors_controller.rb (hunk 1 above)
Comment (despo) [reply app/controllers/admin/sponsors_controller.rb:48]: Yes as I want to run the migrations first. However I should have removed them from the permitted params now, thanks for pointing that out!

File: app/controllers/contact_preferences_controller.rb (hunk 2 above)
Comment (despo) [reply app/controllers/contact_preferences_controller.rb:7]: No reason. Happy to move it there!

File: app/mailers/contact_mailer.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    @contact = contact
+
+    subject = "You have been added to codebar's sponsors mailing list"
+
+    mail(mail_args(contact, subject, 'no-reply@codebar.io'), &:html)
+  end
+
+  private
+
+  helper do
```
Comment (matyikriszta) [inline app/mailers/contact_mailer.rb:16]: @despo the same method has been already defined in `EmailHelper`, should we perhaps use that here?

Comment (despo) [reply app/mailers/contact_mailer.rb:16]: 👍

Comment (despo) [reply app/mailers/contact_mailer.rb:16]: 👍

File: app/models/contact.rb [hunk 4]
```diff
@@ -1,4 +1,32 @@
...
-  has_many :sponsors, through: :member_contacts
+  belongs_to :sponsor
+
+  validates :name, :surname, :email, presence: true
+
+  before_create :set_token
+  after_commit :subscribe_to_mailinglist, if: proc { |c| c.mailing_list_consent }
+  after_commit :unsubscribe_from_mailinglist, if: proc { |c| !c.mailing_list_consent }
+
+  def subscribe_to_mailinglist
```
Comment (matyikriszta) [inline app/models/contact.rb:11]: @despo since mailing list is two words can we rename this method and the one below to `subscribe_to_mailing_list` & `unsubscribe_from_mailing_list` respectively? Also rename the `mailinglist` method to `mailing_list`

File: app/models/sponsor.rb [hunk 5]
```diff
@@ -11,14 +12,26 @@ class Sponsor < ActiveRecord::Base
 
   has_one :address
   has_many :chapters, through: :workshops
-  has_many :workshop_sponsors
-  has_many :workshops, through: :workshop_sponsors
+  has_many :contacts
```
Comment (matyikriszta) [inline app/models/sponsor.rb:15]: @despo this is where I lost the plot a little bit. Can you elaborate on how `Sponsors`, `Contacts`, `Members` and `MemberContacts` work together in this new system? Is `MemberContact` going to be removed eventually?

File: app/models/contact.rb (hunk 4 above)
Comment (despo) [reply app/models/contact.rb:11]: 👍

File: app/models/sponsor.rb (hunk 5 above)
Comment (despo) [reply app/models/sponsor.rb:15]: There was a `Contact` model that had been created when the Sponsor-Contact functionality was originaly implemented but was never utilised. Instead, a member_contact association had been added and used, so up to now, a sponsor had many members as contacts, and could also have the details of a single contact stored directly on the Sponsor model using the `contact_first_name`, `contact_surname` and `email` fields.

I decided to completely drop the Sponsor/Model association (SponsorContacts) and instead use the existing Contact model to enable adding multiple independent contacts, as there is no benefit or need to have members associated with a Sponsor in the app.

After this PR is in I will be completely removing the MemberContact model, but at the moment I still need the association in place in order to correctly migrate existing data using the `sponsors:members:migrate_to_contacts` rake task.

So in short:
In this new functionality a `Sponsor` can have multiple `Contacts`. Nothing else!

Comment (matyikriszta) [reply app/models/sponsor.rb:15]: @despo ok cool, it's what I thought, thanks for clarifying :)

File: app/models/sponsor.rb [hunk 6]
```diff
@@ -11,14 +12,26 @@ class Sponsor < ActiveRecord::Base
...
+  has_many :meetings, -> { order(date_and_time: :desc) }, foreign_key: 'venue_id', inverse_of: :venue
   has_many :member_contacts
-  has_many :contacts, through: :member_contacts, class_name: 'Member', foreign_key: 'member_id'
+  has_many :members, through: :member_contacts
+  has_many :sponsorships, -> { includes([:event]).joins(:event).order('events.date_and_time desc') },
```
Comment (matyikriszta) [inline app/models/sponsor.rb:19]: @despo I'm wondering if the 'name' of this assoc could allude that these are event sponsorships. It would make the `sponsorships_count` method in `SponsorPresenter` clearer.

Comment (despo) [reply app/models/sponsor.rb:19]: Good point. Even though I do think that of the sponsor association names this is the best, it is indeed on clear that it's related to Events. I would like to eventually rename this to event_sponsorships, and apply the same approach to workshop_sponsors and meetings_sponsors, so that we have workshop_sponsorships and meeting_sponsorships.

For now I will rename only the association to event_sponsorships and also add a `has_many :events` on the sponsor model, that I can use to calculate `sponsorship_count` on the presenter.

Comment (matyikriszta) [reply app/models/sponsor.rb]: @despo yep, sounds good

**Review by matyikriszta — APPROVED** [review]
I think that's a 👍 from me
